README.md
# PHP Terminal style
[![Hex.pm](https://img.shields.io/hexpm/l/plug.svg?maxAge=2592000&style=flat-square)](https://github.com/bvanhoekelen/terminal-style/blob/master/LICENSE)
[![Packagist Prerelease](https://img.shields.io/packagist/vpre/bvanhoekelen/terminal-style.svg?style=flat-square)](https://packagist.org/packages/bvanhoekelen/terminal-style)
[![Packagist](https://img.shields.io/packagist/dt/bvanhoekelen/terminal-style.svg?style=flat-square)](https://packagist.org/packages/bvanhoekelen/terminal-style)
[![Github issues](https://img.shields.io/github/issues/bvanhoekelen/terminal-style.svg?style=flat-square)](https://github.com/bvanhoekelen/terminal-style/issues)

## Highlight
- The easiest way to style your text in the command line / terminal
- Change text color to red, green, yellow ...  » [Text color](#text-color)
- Change background color to red, green, yellow ...  » [Background color](#background-color)
- Change background in customer 8-bit color  » [Custom background color](https://github.com/bvanhoekelen/terminal-style/wiki/style-%C2%BB-custom-background)
- Change text style to bold, dim, underlined, blink ...  » [Text style](#text-style)
- Support for Laravel framework » [Laravel](https://laravel.com)
- Easy to install » [installation](#installation)
- Love feedback » [backlog](https://github.com/bvanhoekelen/terminal-style/blob/master/BACKLOG.md) or [create issues](https://github.com/bvanhoekelen/terminal-style/issues)

## How to use
```php
 echo terminal_style($message = null, $color = null, $background = null, $style = null);
```

## Text color
<p align="center"><img src="/assets/terminal-text-color.png" alt="PHP Terminal style set text color" /></p>

### Code example
```php
// Print red text 
echo terminal_style('Here your text', 'red');
```
### Text color names
Use can use: `default`, `yellow`, `red`, `green`, `light-gray` [...](https://github.com/bvanhoekelen/terminal-style/wiki/style-%C2%BB-text-color-names)

See [the full color name list here](https://github.com/bvanhoekelen/terminal-style/wiki/style-%C2%BB-text-color-names)

## Background color
<p align="center"><img src="/assets/terminal-background-color.png" alt="PHP Terminal style set background color" /></p>

### Code example
```php
// Print text with background red
echo terminal_style('Here your text', null, 'red');
```
### Background colors names
Use can use: `default`, `yellow`, `red`, `green`, `light-gray` [...](https://github.com/bvanhoekelen/terminal-style/wiki/style-%C2%BB-background-color-names)

See [the full background color name list here](https://github.com/bvanhoekelen/terminal-style/wiki/style-%C2%BB-background-color-names) or see [custom background color](https://github.com/bvanhoekelen/terminal-style/wiki/style-%C2%BB-custom-background)

## Text style
<p align="center"><img src="/assets/terminal-text-style.png" alt="PHP Terminal style set style" /></p>

### Code example
```php
// Print text style bold
echo terminal_style('Here your text', null, null, 'bold');
```
### Text styles

Use can use: `default`, `bold `, `dim`, `italic`, `underlined`, 'blink' [...](https://github.com/bvanhoekelen/terminal-style/wiki/style-%C2%BB-text-style-names)

See [the full text style name list here](https://github.com/bvanhoekelen/terminal-style/wiki/style-%C2%BB-text-style-names)

# Help, docs and links
- [Wiki](https://github.com/bvanhoekelen/terminal-style/wiki)
- [Backlog](https://github.com/bvanhoekelen/terminal-style/blob/master/BACKLOG.md)
- [Change log](https://github.com/bvanhoekelen/terminal-style/blob/master/CHANGELOG.md)
- [Packagist](https://packagist.org/packages/bvanhoekelen/terminal-style)

## Backlog & Feedback
If you have any suggestions to improve this php terminal style tool? Please add your feature, bug or improvement to the [BACKLOG.dm](https://github.com/bvanhoekelen/terminal-style/blob/master/BACKLOG.md). Or create a [issues](https://github.com/bvanhoekelen/terminal-style/issues).
- [Open backlog](https://github.com/bvanhoekelen/terminal-style/blob/master/BACKLOG.md)
- [Create issues](https://github.com/bvanhoekelen/terminal-style/issues)


# Installation

## Install with Laravel
Get PHP terminal style tool by running the Composer command in the command line. 
```{r, engine='bash', count_lines}
 $ composer require bvanhoekelen/terminal-style
```

Open your file and use `terminal_style()`
```php
// Print red text 
echo terminal_style('Here your text', 'red');

```

## Install with composer
Get PHP terminal style tool by running the Composer command in the command line. 
```{r, engine='bash', count_lines}
 $ composer require bvanhoekelen/terminal-style
```

Open your file and use `terminal_style()`
```php
// Require vender autoload
require_once('../vendor/autoload.php');

// Print red text 
echo terminal_style('Here your text', 'red');

```

## Overview
<p align="center"><img src="/assets/terminal-all-styles.png" alt="PHP Terminal style all styles" /></p>

## Custom background
<p align="center"><img src="/assets/terminal-custom-background.png" alt="PHP Terminal style custom backgroud" /></p>

See [custom background color](https://github.com/bvanhoekelen/terminal-style/wiki/style-%C2%BB-custom-background)

CHANGELOG.md
# Change Log
## [v1.0.5](https://github.com/bvanhoekelen/terminal-style/tree/v1.0.4) - 2020-01-31
- Support PHP 7.4, fix implode order issues. See #2.

## [v1.0.4](https://github.com/bvanhoekelen/terminal-style/tree/v1.0.4) - 2018-02-19
- Fix composer validate error

## [v1.0.3](https://github.com/bvanhoekelen/terminal-style/tree/v1.0.3) - 2017-11-20
- Remove larapack/dd dependency
- Change dd() to die()

## [v1.0.2](https://github.com/bvanhoekelen/terminal-style/tree/v1.0.2) - 2017-05-18
- Change error message
- Add custom background color
- Disable color for windows and web
- Add colors `primary`, `success`, `info`, `warning` and `danger`

## [v1.0.1](https://github.com/bvanhoekelen/terminal-style/tree/v1.0.1) - 2017-04-26
- Massage can be empty or null

## [v1.0.0](https://github.com/bvanhoekelen/terminal-style/tree/v1.0.0) - 2017-04-20
- Start project on Packagist


src/terminal-style.php
<?php

declare(strict_types=1);

use src\Configuration\StyleConfigurations;
use src\Service\StyleAppenderService;

if (!function_exists('terminal_style')) {

    /**
     * @param null $message
     * @param null $color
     * @param null $background
     * @param null $style
     * @return string
     */
    function terminal_style($message = null, $color = null, $background = null, $style = null): string
    {
        if ($message === null) {
            return '';
        }

        $message = is_null($message) ? '' : $message;

        // Only for terminal
        if (php_sapi_name() !== "cli") {
            return $message;
        }

        // Only for linux not for windows (PowerShell)
        if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
            return $message;
        }

        // Detect custom background mode
        if (is_int($color) and $color >= 16) {
            $background = 5;
            $style = 48;
        }

        // Set default
        $code = [];

        $service = new StyleAppenderService();
        $code = $service->appendStyle($code, $background, StyleConfigurations::BACKGROUND_COLOUR_CODES, 'background');
        $code = $service->appendStyle($code, $color, StyleConfigurations::TEXT_COLOUR_CODES, 'text');
        $code = $service->appendStyle($code, $style, StyleConfigurations::STYLE_CODES, 'style');

        // Set background
        return "\e[" . implode(';', $code) . "m" . $message . "\e[0m";
    }
}


examples/04_show_all.php
<?php

require_once('../vendor/autoload.php');

echo PHP_EOL . " Show message in color" . PHP_EOL . PHP_EOL;
echo str_repeat('-', 40) . PHP_EOL . PHP_EOL;

// Show colors
$colors = [
    'default',
    'white',
    'black',
    'red',
    'green',
    'yellow',
    'blue',
    'magenta',
    'cyan',
    'gray',
    'dark-gray',
    'light-gray',
    'light-red',
    'light-green',
    'light-yellow',
    'light-blue',
    'light-magenta',
    'light-cyan'
];

$textStyle = [
    'default',
    'bold',
    'dim',
    'italic',
    'underlined',
    'blink',
    'reverse',
    'hidden',
    'password'
];

$pad = 37;

foreach ($colors as $color)
{
    echo str_pad(" > Text color " . $color, $pad). " : " . terminal_style('This is ' . $color, $color) . PHP_EOL;
}

echo PHP_EOL;
echo PHP_EOL . " Show message in background color" . PHP_EOL . PHP_EOL;
echo str_repeat('-', 40) . PHP_EOL . PHP_EOL;

// Show colors
foreach ($colors as $color)
{
    echo str_pad(" > Background color " . $color, $pad). " : " . terminal_style('This is ' . $color, null, $color) . PHP_EOL;
}

echo PHP_EOL;
echo PHP_EOL . " Change message style" . PHP_EOL . PHP_EOL;
echo str_repeat('-', 40) . PHP_EOL . PHP_EOL;

// Show text style
foreach ($textStyle as $style)
{
    echo str_pad(" > Background color " . $style, $pad). " : " . terminal_style('this is ' . $style, null, null, $style) . PHP_EOL;
}

src/Service/StyleAppenderService.php
<?php

declare(strict_types=1);

namespace src\Service;

/**
 * Class StyleAppenderService
 * @package src\Service
 */
class StyleAppenderService
{
    /**
     * @var array
     */
    protected $styleMessages = [
        'background' => ' > terminal_style(): Background color "%s" does not exist. You can only use the background colors above',
        'text' => ' > terminal_style(): Text color "%s" does not exist. You can only use the following text colors',
        'style' => ' > terminal_style(): Text style "%s" does not exist. You can only use the text styles above',
    ];

    /**
     * Appends the style according to the type
     *
     * @param array $code Collection of codes for styling
     * @param int|string|null $color Color. Can either be string or an integer or null
     * @param array $configuration Configuration array (@see StyleConfigurations)
     * @param string $type The type of style is currently added to the code array
     *
     * @return array
     */
    public function appendStyle(array $code, $color, array $configuration, string $type): array
    {
        $color = (!$color) ? 'default' : $color;

        // Set text color
        if (\is_int($color))
            $code[] = $color;
        elseif (isset($configuration[$color]))
            $code[] = $configuration[$color];
        else {
            $this->printErrorMessage($type, $color, $configuration);
        }

        return $code;
    }

    /**
     * @param $type
     * @param $style
     * @param $styleCollection
     */
    protected function printErrorMessage($type, $style, $styleCollection): void
    {
        \print_r(\array_keys($styleCollection));
        die(\sprintf($this->styleMessages[$type], $style));
    }
}

src/Configuration/StyleConfigurations.php
<?php

declare(strict_types=1);

namespace src\Configuration;

/**
 * Class StyleConfigurations
 * @package src\Configuration
 */
final class StyleConfigurations
{
    const TEXT_COLOUR_CODES = [
        // Label
        'default' => 39,
        'primary' => 34,
        'success' => 32,
        'info' => 36,
        'warning' => 33,
        'danger' => 31,

        // Colors
        'white' => 97,
        'black' => 30,
        'red' => 31,
        'green' => 32,
        'yellow' => 33,
        'blue' => 34,
        'magenta' => 35,
        'cyan' => 36,
        'gray' => 37,

        // Light colors
        'light-gray' => 37,
        'light-red' => 91,
        'light-green' => 92,
        'light-yellow' => 93,
        'light-blue' => 94,
        'light-magenta' => 95,
        'light-cyan' => 96,

        // Dark colors
        'dark-gray' => 90,
    ];

    const BACKGROUND_COLOUR_CODES = [
        // Label
        'default' => 39,
        'primary' => 44,
        'success' => 42,
        'info' => 46,
        'warning' => 43,
        'danger' => 41,

        // Colors
        'white' => 39,
        'black' => 40,
        'red' => 41,
        'green' => 42,
        'yellow' => 43,
        'blue' => 44,
        'magenta' => 45,
        'cyan' => 46,
        'gray' => 47,

        // Light colors
        'light-gray' => 47,
        'light-red' => 101,
        'light-green' => 102,
        'light-yellow' => 103,
        'light-blue' => 104,
        'light-magenta' => 105,
        'light-cyan' => 106,

        // Dark colors
        'dark-gray' => 100,
    ];

    const STYLE_CODES = [
        'default' => 0,
        'bold' => 1,
        'dim' => 2,
        'italic' => 3,
        'underlined' => 4,
        'blink' => 5,
        'reverse' => 7,
        'hidden' => 8,
        'password' => 8,
    ];
}